### PR TITLE
Use config files for operator and PAF client node

### PR DESCRIPTION
**Pull request by OlivierChirouze**

### CODE REVIEW

File: paf-mvp-demo-express/src/websites/portal.ts [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  host: 'portal.onekey.network',
+};
+
+export const portalWebSiteApp = new App(name).setHostName(host);
+
+(async () => {
+  const keyStore = new PublicKeyStore(s2sOptions);
+
+  // Little trick here, we use an OperatorClient object
+  const { config, currentPrivateKey } = await parseConfig<ClientNodeConfig>('configs/portal-client/config.json');
```
Comment (OlivierChirouze) [inline paf-mvp-demo-express/src/websites/portal.ts:51]: This file didn't change apart from here, it reads the JSON file to get config

File: paf-mvp-operator-client-express/examples/example-client-node.ts [hunk 2]
```diff
@@ -1,24 +1,24 @@
-import { addClientNodeEndpoints } from '@operator-client/client-node';
-import express from 'express';
+import { ClientNode } from '@operator-client/client-node';
 import { getTimeStampInSec } from '@core/timestamp';
 
 // This is just an example of a basic client node configuration
-addClientNodeEndpoints(
-  express(),
+export const pofPublisherClientNode = new ClientNode(
```
Comment (OlivierChirouze) [inline paf-mvp-operator-client-express/examples/example-client-node.ts:5]: to be updated in a future PR

File: paf-mvp-operator-express/examples/example-operator-api.ts [hunk 3]
```diff
@@ -1,24 +1,24 @@
-import { addOperatorApi, Permission } from '@operator/operator-api';
-import express from 'express';
 import { getTimeStampInSec } from '@core/timestamp';
+import { OperatorNode, Permission } from '@operator/operator-node';
 
 // This is just an example of a basic operator node configuration
-addOperatorApi(
-  express(),
+export const operator = new OperatorNode(
```
Comment (OlivierChirouze) [inline paf-mvp-operator-express/examples/example-operator-api.ts:5]: To be updated in a future PR

**Comment by OlivierChirouze** [comment]
Next step: make the PAF client node accept multiple JSON files

File: paf-mvp-demo-express/src/crto1-operator.ts [hunk 4]
```diff
@@ -34,11 +33,8 @@ ZxbtbfH3C+VfhheolRApHZzSW96pUOPiHA7SRNkO41FSGDGTiKvBXd/P
   privacyPolicyUrl: 'https://crto-poc-1.onekey.network/privacy',
 };
 
-export const crtoOneOperatorApp = new App(crtoOneOperatorConfig.name).setHostName(crtoOneOperatorConfig.host);
-
 // This host supports the Operator API
-addOperatorApi(
-  crtoOneOperatorApp.app,
+export const crtoOneOperatorApp = new OperatorNode(
```
Comment (Aerion) [inline paf-mvp-demo-express/src/crto1-operator.ts]: nit: wdyt of renaming it to `crtoOneOperator`?

File: paf-mvp-operator-express/src/operator-node.ts [hunk 5]
```diff
@@ -56,14 +59,20 @@ export enum Permission {
   WRITE = 'WRITE',
 }
 
-export type AllowedDomains = { [domain: string]: Permission[] };
+export type AllowedHosts = { [host: string]: Permission[] };
+
+export interface OperatorConfig {
```
Comment (Aerion) [inline paf-mvp-operator-express/src/operator-node.ts:62]: WDYT of renaming this to `OperatorNodeConfig`, given that there is `ClientNodeConfig` for the `ClientNode` class, and that it's for the `OperatorNode` class?

File: paf-mvp-demo-express/src/index.ts [hunk 6]
```diff
@@ -28,6 +28,12 @@ const relative = (dir: string) => join(__dirname, dir);
               res.setHeader('Access-Control-Allow-Origin', '*');
               res.setHeader('Cache-Control', 'public, max-age=604800');
             }
+            if (/(js)$/.test(filePath)) {
```
Comment (Aerion) [inline paf-mvp-demo-express/src/index.ts:31]: FAR: You already have it in mind, but to put in writing, the regex should test for the file extension, i.e. including the dot `/\.js$/`

**Review by Aerion — COMMENTED** [review]
Looks good overall, I put minor comments.
Though as it's a big PR, I'll let another maintainer review it :)

File: paf-mvp-core-js/src/express/config.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+/**
+ * Extract the config object, the Identity object, and the current private key from a config JSON file
+ * @param configPath
+ */
+export const parseConfig = async <T extends ClientNodeConfig | OperatorConfig>(
+  configPath: string
+): Promise<Parsed<T>> => {
+  const config = JSON.parse((await fs.promises.readFile(configPath)).toString()) as T;
```
Comment (RomainLofaso) [inline paf-mvp-core-js/src/express/config.ts:59]: Since JSON-Schema is on the project, I am wondering if it is useful to use here for its initial purpose: verifying data before deserializing it.

File: paf-mvp-core-js/src/express/utils.ts [hunk 8]
```diff
@@ -79,17 +79,12 @@ export const getPayload = <T>(req: Request): T => {
 /**
  * Escape a string to be used in a regular expression.
  * Stolen from https://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex#answer-6969486
+ * Examples:
+ *  somewebsite.com/testURL?key=val#anchor => somewebsite\\.com/testURL\\?key=val#anchor
```
Comment (RomainLofaso) [inline paf-mvp-core-js/src/express/utils.ts:83]: Thank you!

File: paf-mvp-demo-express/src/websites/portal.ts [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+    }
+  });
+
+  const requestWithoutBodyVerifier = (request: RequestWithContext<MessageBase>) =>
+    new RequestVerifier(keyStore.provider, new RequestWithoutBodyDefinition()).verifySignature(request);
+  const postIdsPrefsRequestVerifier = (request: RequestWithContext<PostIdsPrefsRequest>) =>
+    new RequestVerifier(keyStore.provider, new RequestWithBodyDefinition()).verifySignature(request);
+  const responseVerifier = (response: ResponseType) =>
+    new ResponseVerifier(keyStore.provider, new ResponseDefinition()).verifySignature(response);
+  /*
```
Comment (RomainLofaso) [inline paf-mvp-demo-express/src/websites/portal.ts:161]: Commented code. 

If it is meaningful to keep it, maybe a bit of explanation inside it.

File: paf-mvp-core-js/src/express/config.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+      publicKey: (await fs.promises.readFile(relative(keyPair.publicKeyPath))).toString(),
+      privateKey: (await fs.promises.readFile(relative(keyPair.privateKeyPath))).toString(),
+      start: getTimeStampInSec(new Date(keyPair.startDateTimeISOString)),
+      end: keyPair.endDateTimeISOString ? getTimeStampInSec(new Date(keyPair.endDateTimeISOString)) : undefined,
+    }))
+  );
+};
+
+export interface Parsed<T extends ClientNodeConfig | OperatorConfig> {
+  config: T;
```
[the remark this answers is not in the document]
Comment (OlivierChirouze) [reply paf-mvp-core-js/src/express/config.ts:47]: I agree with you. Returning the original "raw" data was convenient originally, but it's not clean design.
However, I still need a generic to differentiate the two types of config.

[the remark this answers is not in the document]
Comment (Aerion) [reply paf-mvp-core-js/src/express/config.ts:47]: btw this isn't a blocker, the current solution is fine as-is.

Would having a return type be a `INodeConfig`, with two implementations (`ClientNodeConfig` and `OperatorNodeConfig`), be a solution? Or just having a return type `ClientNodeConfig | OperatorNodeConfig`?
Where `ClientNodeConfig` and `OperatorNodeConfig` would be processed versions of the config.

[the remark this answers is not in the document]
Comment (OlivierChirouze) [reply paf-mvp-core-js/src/express/config.ts:47]: Yes, I'm working on the second solution, thanks ;-)

[the remark this answers is not in the document]
Comment (OlivierChirouze) [reply paf-mvp-core-js/src/express/config.ts]: Fixed.

File: paf-mvp-operator-express/src/operator-node.ts [hunk 11]
```diff
@@ -0,0 +1,423 @@
+// Expiration: now + 3 months
```
[the remark this answers is not in the document]
Comment (OlivierChirouze) [reply paf-mvp-operator-express/src/operator-node.ts]: thanks!

File: paf-mvp-operator-express/src/operator-node.ts (hunk 5 above)
Comment (OlivierChirouze) [reply paf-mvp-operator-express/src/operator-node.ts]: thanks, done

**Review by OlivierChirouze — COMMENTED** [review]
There are a couple of other comments I will handle in another commit

File: paf-mvp-core-js/src/express/express-apps.ts [hunk 12]
```diff
@@ -6,23 +6,21 @@ import vhost from 'vhost';
 /**
  * Encapsulate an Express App that listens on the local IP
  */
-export class MainApp {
-  constructor(public app = express()) {
-    addMiddlewares(this.app);
-  }
+export class App {
+  public hostName: string | undefined;
```
[the remark this answers is not in the document]
Comment (OlivierChirouze) [reply paf-mvp-core-js/src/express/express-apps.ts:10]: Actually, there is one app that doesn't have a hostname. I have extracted it as a dedicated class, please review it, thanks!

[the remark this answers is not in the document]
Comment (Aerion) [reply paf-mvp-core-js/src/express/express-apps.ts]: Nice, thanks!

File: paf-mvp-core-js/src/express/config.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+  configPath: string,
+  identity: IdentityConfig
+): Promise<{ publicKey: string; privateKey: string; start: number; end: number }[]> => {
+  // Key paths are relative to the config file
+  const relative = (dir: string) => join(path.dirname(configPath), dir);
+
+  // Add start and end timestamp in seconds, and load key values
+  return await Promise.all(
+    identity.keyPairs.map(async (keyPair) => ({
+      publicKey: (await fs.promises.readFile(relative(keyPair.publicKeyPath))).toString(),
```
[the remark this answers is not in the document]
Comment (OlivierChirouze) [reply paf-mvp-core-js/src/express/config.ts]: indeed.
I added unit tests to make it tested and somehow documented. I believe it's ok to return only the first error. I don't think it's worth investigating too much for this "edge case", WDYT?

File: paf-mvp-core-js/src/express/config.ts (hunk 7 above)
Comment (OlivierChirouze) [reply paf-mvp-core-js/src/express/config.ts]: This is an excellent point.
Since I think it's a bit of work worth, I suggest to keep it FAR. WDYT?

Comment (RomainLofaso) [reply paf-mvp-core-js/src/express/config.ts]: Agree.

File: paf-mvp-core-js/src/express/config.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (RomainLofaso) [reply paf-mvp-core-js/src/express/config.ts]: ok. Move forward.
